### PR TITLE
Create data provider for an ideal components bag

### LINKED ISSUE #1318: Prototype an initial data provider for the Ideal Components Bag
This is a subtask of #1317.

Prototype and land an initial design for the provider data. This first pass does not need to be perfect, but will inform what additional information we need from CLDR. This step will probably use some of the skeleton matching and manipulation already present in the current implementation.

### PR DESCRIPTION
**Pull request by jedel1043**
Closes #1318

This follows the [Provider data format](https://docs.google.com/document/d/18v9fQcDvHDkG_7Hx6rDt1r3Mq6_JMecgORgOH4yXAWU/edit#heading=h.yoca0tmipd40) JSON as closely as possible.

There's a lot of work to be done still, but I'm opening a draft to seek for some comments on the structure of the provider.

Some thoughts:

- Would be good to have a concept of a generic `Pattern` (not to be confused with `GenericPattern`) restricted to a subset of `FieldSymbol`s. That way, e.g. `year_month_day` could be restricted to a `Pattern` with only the `Year`, `Month` and `Day` symbols, which should make parsing a bit easier.
- `preferred_hour_cycle` is already provided by the current version of CLDR, right?
- `time_zone` is still a WIP, since I'm not sure what should be the underlying data needed for that.
- Is there a need for `glue` in time formats?
- Always open for better name proposals 😅
- The proposed JSON includes "mixed" patterns with format fields + placeholders. I'm creating an entirely new pattern struct to handle this, but there's a possibility of unifying `Pattern`, `GenericPattern` and this new `MixedPattern` in a single struct.

### CODE REVIEW

File: experimental/datetime_bag_v2/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,37 @@
+[package]
+name = "datetime_bag_v2"
```
Comment (Manishearth) [inline experimental/datetime_bag_v2/Cargo.toml:2]: @sffc @nordzilla ought this be a separate crate or just an experimental module in DTF?

Comment (jedel1043) [reply experimental/datetime_bag_v2/Cargo.toml:2]: Made a new crate based on [this](https://github.com/unicode-org/icu4x/issues/1318#issuecomment-1040752383) comment, but that was several months ago.

Comment (Manishearth) [reply experimental/datetime_bag_v2/Cargo.toml:2]: Ah, okay. I wasn't super plugged in to that part of that decision so I'll defer to the folks who are

Comment (sffc) [reply experimental/datetime_bag_v2/Cargo.toml:2]: I stand by my suggestion from February:

> The lion's share of the code in the datetime crate is built on assumptions of the old data model, and many of the pieces independent of the data model have been moved to the calendar crate. I therefore believe the best place to start working on this is in a new experimental crate. At the end, once things are working, we will reconcile the two.

Comment (Manishearth) [reply experimental/datetime_bag_v2/Cargo.toml:2]: Works for me!

**Review by sffc — COMMENTED** [review]
Praise: This seems like a good data model.

Suggestion: Maybe do this in a way that we can use MixedPatternItem also for regular PatternItem so we don't have so much code duplication. PatternItem doesn't need `Placeholder(u8)`. If it encounters one at runtime, it can either error out or substitute some other token in its place.

CC @zbrankecki who has thought a lot about these things.

**Comment by jedel1043** [comment]
cc @zbraniecki

**Comment by jedel1043** [comment]
> Suggestion: Maybe do this in a way that we can use MixedPatternItem also for regular PatternItem so we don't have so much code duplication. PatternItem doesn't need `Placeholder(u8)`. If it encounters one at runtime, it can either error out or substitute some other token in its place.

I think `MixedPatternItem` is more akin to `GenericPatternItem` than to `PatternItem`, since it should eventually support something like the `GenericPattern::combined` operation. Also, I have some ideas on how to simplify some of the code to reduce duplication :)

**Pull request closed without merging**